<launch>
  <master auto="start"/>
  <param name="/use_sim_time" value="false"/>
  <arg name="use_map_topic" default="true"/>
  <arg name="scan_topic" default="wc_scan"/>
  <arg name="no_static_map" default="false"/>
  <arg name="base_global_planner" default="navfn/NavfnROS"/>
  <arg name="base_local_planner" default="dwa_local_planner/DWAPlannerROS"/>
  <arg name="map" value="$(find goto_odroid)/config/map/UTS_map.yaml"/>

  <!-- Run tf -->
  <!--node pkg="tf" type="static_transform_publisher" name="bl_laser" args="-0.23 -0.2 0 0 0 0 /base_link /laser 40" /--> <!-- laser highest position -->
  <node pkg="tf" type="static_transform_publisher" name="bl_laser" args=" 0.67 0.32 0 1.5708 0 3.1416 /base_link /laser 40" /> <!-- laser around the foot --> 
  <node pkg="tf" type="static_transform_publisher" name="bf_bl" args="0.0 0 0 0.0 0.0 0.0 /base_footprint /base_link 40" />
  <node pkg="tf" type="static_transform_publisher" name="refl_pole" args="-0.29 -0.2 0 0 0 0 /base_link /pole 40" />

  <node pkg="goto_odroid" type="convert_scan" name="convert_scan" output="screen"/>
  <node pkg="laser_box_measurement" type="transform_measurement" name="transform_measurement" output="screen"/>

  <!-- Run the map server -->
  <node pkg="map_server" type="map_server" name="map_server" args="$(arg map)"/>

  <!-- Run AMCL -->
  <node pkg="amcl" type="amcl" respawn="true" name="amcl">
	<param name="use_map_topic" value="$(arg use_map_topic)"/>
  	
	<!-- Overall filter parameters -->
	<param name="min_particles" value="100"/>
	<param name="max_particles" value="1000"/>
	<param name="kld_err" value="0.05"/> 
	<param name="kld_z" value="0.95"/>
	<param name="update_min_d" value="0.01"/>
	<param name="update_min_a" value="0.01"/>
	<param name="resample_interval" value="1"/>
	<param name="transform_tolerance" value="0.1"/>
	<!--param name="recovery_alpha_slow" value="0.001"/-->
	<!--param name="recovery_alpha_fast" value="0.1"/-->
	<param name="gui_publish_rate" value="10.0"/>

	<!-- Laser model parameters -->
	<param name="laser_max_beams" value="120"/>
	<param name="laser_z_hit" value="0.95"/>
	<param name="laser_z_short" value="0.01"/>
	<param name="laser_z_max" value="0.05"/>
	<param name="laser_z_rand" value="0.5"/>
	<param name="laser_sigma_hit" value="0.2"/>
	<param name="laser_lambda_short" value="0.1"/>
    <param name="laser_likelihood_max_dist" value="1.0"/>
	<param name="laser_model_type" value="likelihood_field"/>

	<!-- Odometry model parameters -->
  	<param name="odom_model_type" value="diff"/>
	<param name="odom_alpha1" value="1.5"/>
	<param name="odom_alpha2" value="1.5"/>
	<param name="odom_alpha3" value="1.5"/>
	<param name="odom_alpha4" value="1.5"/>
	<param name="odom_frame_id" value="odom"/>
	<param name="base_frame_id" value="base_link"/>
	<param name="global_frame_id" value="map"/>
	<param name="tf_broadcast" value="true"/>

	<remap from="scan" to="$(arg scan_topic)"/>
  </node>

  <!-- Run scan matcher -->
  <node pkg="laser_scan_matcher" type="laser_scan_matcher_node" name="laser_scan_matcher_node" output="screen">
	<param name="fixed_frame" value= "/odom"/>
	<param name="base_frame" value="/base_footprint"/>

	<param name="use_imu" value="false"/>
	<param name="use_odom" value="false"/>
	<param name="use_vel" value="false"/>

	<param name="use_cloud_input" value="false"/>
	<param name="cloud_range_min" value="0.1"/>
	<param name="cloud_range_max" value="50.0"/>

	<param name="kf_dist_linear" value="0.01"/>
	<param name="kf_dist_angular" value="0.0175"/>

	<param name="publish_tf" value="true"/>
	<param name="publish_pose" value="true"/>
	<param name="publish_pose_stamped" value="true"/>

	<param name="max_iterations" value="20"/>
	<param name="max_correspondence_dist" value="0.3"/>
	<param name="max_angular_correction_deg" value="40"/>
	<param name="max_linear_correction" value="0.50"/>
	<param name="epsilon_xy" value="0.000001"/>
	<param name="epsilon_theta" value="0.000001"/>
	<param name="outliers_maxPrec" value="0.85"/>
	
	<param name="sigma" value="0.05"/>
	<param name="use_corr_tricks" value="1"/>
	<param name="restart" value="1"/>
	<param name="restart_threshold_mean_error" value="0.4"/>
	<param name="restart_dt" value="1.0"/>
	<param name="restart_dtheta" value="0.1"/>
	<param name="clustering_threshold" value="0.25"/>
	<param name="orientation_neighbourhood" value="10"/>
	<param name="use_point_to_line_distance" value="1"/>
	<param name="do_alpha_test" value="0"/>
	<param name="do_alpha_test_thresholdDeg" value="20.0"/>
	<param name="outliers_adaptive_order" value="0.7"/>
	<param name="outliers_adaptive_mul" value="2.0"/>
	<param name="do_visibility_test" value="0"/>
	<param name="outliers_remove_doubles" value="1"/>
	<param name="do_compute_covariance" value="0"/>
	<param name="debug_verify_trick" value="0"/>
	<param name="use_ml_weights" value="0"/>
	<param name="use_sigma_weights" value="0"/>

	<remap from="scan" to="$(arg scan_topic)"/>
  </node>

  <!-- Run move base -->
  <node pkg="move_base" type="move_base" respawn="true" name="move_base" output="screen">
	<param name="base_global_planner" value="$(arg base_global_planner)"/>
	<param name="base_local_planner" value="$(arg base_local_planner)"/>
	<rosparam file="$(find goto_odroid)/config/planner/base_local_planner.yaml" command="load" />

	<!-- observation sources located in costmap_common_parameters.yaml -->
    	<rosparam file="$(find goto_odroid)/config/planner/costmap_common.yaml" command="load" ns="global_costmap" />
    	<rosparam file="$(find goto_odroid)/config/planner/costmap_common.yaml" command="load" ns="local_costmap" />

	<!-- local costmap, need size -->
    	<rosparam file="$(find goto_odroid)/config/planner/local_costmap.yaml" command="load" ns="local_costmap" />
	<param name="local_costmap/width" value="10.0"/>
	<param name="local_costmap/height" value="10.0"/>
	
	<!-- Static global costmap, static map provide size -->
    	<rosparam file="$(find goto_odroid)/config/planner/global_costmap.yaml" command="load" ns="global_costmap" unless="$(arg no_static_map)"/>
  </node>

</launch>
